### PR TITLE
Handling feedback on utterance level

### LINKED ISSUE #219: Handling feedback on utterance level
We could have `self._feedbacks: List[Feedback] = []` on the dialogue level. 

Feedback may be added for all utterances in the given dialogue or none. It is added independently for a single utterance. This way it is also possible to add feedback to previous utterances, not only to the current one (useful especially if feedback is just a "like" added for a message in a web chat). 

Then we could have `add_feedback` in `Dialogue` and `register_user_feedback` in DialogueConnector.

### PR DESCRIPTION
**Pull request by WerLaj**
Fixes #219

### CODE REVIEW

**Review by NoB0 — APPROVED** [review]
LGTM.

File: dialoguekit/connector/dialogue_connector.py [hunk 1]
```diff
@@ -115,6 +116,16 @@ def register_agent_utterance(
         else:
             self._user.receive_utterance(annotated_utterance)
 
+    def register_user_feedback(
+        self, utterance_feedback: UtteranceFeedback
+    ) -> None:
+        """Registers users feedback for a utterance from dialogue history.
```
[the remark this answers is not in the document]
Comment (WerLaj) [reply dialoguekit/connector/dialogue_connector.py]: Fixed

File: dialoguekit/core/dialogue.py [hunk 2]
```diff
@@ -103,6 +110,23 @@ def add_utterance(self, utterance: Utterance) -> None:
...
+
+        Args:
+            utterance_feedback: User's feedback.
+        """
+        self._utterance_feedbacks = [
+            feedback
+            for feedback in self._utterance_feedbacks
+            if feedback.utterance_id != utterance_feedback.utterance_id
+        ]
+        self._utterance_feedbacks.append(utterance_feedback)
```
Comment (kbalog) [inline dialoguekit/core/dialogue.py:128]: This is not very efficient, nor elegant. Why not store feedback as a dict, indexed by utterance_id? Then it's enough to set `self._utterance_feedbacks[utterance_feedback.utterance_id] = utterance_feedback`.

File: dialoguekit/core/dialogue.py [hunk 3]
```diff
@@ -77,6 +79,11 @@ def utterances(self) -> List[Utterance]:
         """Returns the utterances in the dialogue."""
         return self._utterances
 
+    @property
+    def utterance_feedbacks(self) -> List[UtteranceFeedback]:
+        """Returns feedbacks on utterances in the dialogue."""
+        return self._utterance_feedbacks
+
```
Comment (kbalog) [inline dialoguekit/core/dialogue.py:86]: Is this ever going to be accessed this way? I'd assume that we either want to fetch the feedback for a given utterance (based on utterance_id) or get the entire dialogue with all utterances and feedbacks.

File: dialoguekit/core/dialogue.py [hunk 4]
```diff
@@ -34,6 +35,7 @@ def __init__(
         else:
             self._conversation_id = conversation_id
         self._utterances: List[Utterance] = []
+        self._utterance_feedbacks: List[UtteranceFeedback] = []
```
Comment (kbalog) [inline dialoguekit/core/dialogue.py:38]: Consider storing these in a dict instead.

File: dialoguekit/connector/dialogue_connector.py [hunk 5]
```diff
@@ -115,6 +116,16 @@ def register_agent_utterance(
         else:
             self._user.receive_utterance(annotated_utterance)
 
+    def register_user_feedback(
+        self, utterance_feedback: UtteranceFeedback
```
Comment (kbalog) [inline dialoguekit/connector/dialogue_connector.py:120]: Do we expect the caller to create an UtteranceFeedback object? Wouldn't it make more sense to take utterance_id and BinaryFeedback as arguments?

**Review by kbalog — CHANGES_REQUESTED** [review]

File: dialoguekit/connector/dialogue_connector.py (hunk 5 above)
Comment (WerLaj) [reply dialoguekit/connector/dialogue_connector.py]: Done

File: dialoguekit/core/dialogue.py (hunk 4 above)
Comment (WerLaj) [reply dialoguekit/core/dialogue.py]: Done

File: dialoguekit/core/dialogue.py (hunk 3 above)
Comment (WerLaj) [reply dialoguekit/core/dialogue.py]: Fixed

File: dialoguekit/core/dialogue.py (hunk 2 above)
Comment (WerLaj) [reply dialoguekit/core/dialogue.py]: Done

File: dialoguekit/connector/dialogue_connector.py [hunk 6]
```diff
@@ -117,14 +117,18 @@ def register_agent_utterance(
             self._user.receive_utterance(annotated_utterance)
 
     def register_user_feedback(
-        self, utterance_feedback: UtteranceFeedback
+        self, feedback: BinaryFeedback, utterance_id: str
```
Comment (kbalog) [inline dialoguekit/connector/dialogue_connector.py:120]: Nit: Mention here that currently only binary feedback is accepted (BinaryFeedback.{NEGATIVE|POSITIVE}).

**Review by kbalog — APPROVED** [review]
LGTM

File: dialoguekit/connector/dialogue_connector.py (hunk 6 above)
Comment (WerLaj) [reply dialoguekit/connector/dialogue_connector.py:120]: Done

**Pull request merged**